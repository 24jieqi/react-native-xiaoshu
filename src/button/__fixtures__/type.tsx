/**
 * title: 按钮类型
 * description: 按钮有五种类型：主按钮（默认）、朦胧按钮、边框按钮、幽灵按钮和链接按钮。
 */

import React, { memo } from 'react'

import {
  Button,
  Card,
  Space,
  Divider,
} from '@fruits-chain/react-native-xiaoshu'

const onPress = () => {
  console.log('点击按钮 => ', new Date().getTime())
}

const ButtonType: React.FC = () => {
  return (
    <Card title="按钮类型" square>
      <Space>
        <Space>
          <Button
            type="primary"
            text="primary 主按钮&点击事件防抖动"
            onPress={onPress}
            onPressDebounceWait={500}
          />
          <Button
            type="hazy"
            text="hazy 朦胧按钮"
            onPress={onPress}
            onPressDebounceWait={300}
          />
          <Button type="outline" text="outline 边框按钮" onPress={onPress} />
          <Button type="ghost" text="ghost 幽灵按钮" onPress={onPress} />
          <Button type="link" text="link 链接按钮" onPress={onPress} />
        </Space>

        <Divider>·</Divider>

        <Space direction="horizontal" wrap>
          <Button type="primary" text="primary" onPress={onPress} />
          <Button type="hazy" text="hazy" onPress={onPress} />
          <Button type="outline" text="outline" onPress={onPress} />
          <Button type="ghost" text="ghost" onPress={onPress} />
          <Button type="link" text="link" onPress={onPress} />
        </Space>
      </Space>
    </Card>
  )
}

export default memo(ButtonType)
