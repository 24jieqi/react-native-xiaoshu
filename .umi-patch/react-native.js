export * from 'react-native-web'

export const InputAccessoryView = () => {}
