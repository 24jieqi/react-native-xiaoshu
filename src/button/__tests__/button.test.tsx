import TOKENS from '@fruits-chain/design-tokens-bailu'
import React from 'react'
import type { ViewStyle } from 'react-native'
import { StyleSheet } from 'react-native'
import type { ReactTestInstance } from 'react-test-renderer'
import { create } from 'react-test-renderer'

import Button from '../'
import { customRender, fireEvent } from '../../__tests__/test-utils'

describe('Button', () => {
  it('render snapshot', () => {
    const tree = create(
      <Button accessibilityLabel="button" type="primary" text="primary" />,
    ).toJSON()

    expect(tree).toMatchSnapshot()
  })

  it('hairline', () => {
    const { getByLabelText, getAllByRole } = customRender(
      <>
        <Button
          accessibilityLabel="button-primary"
          text="primary"
          type="outline"
          hairline
        />
        <Button
          accessibilityLabel="button-primary2"
          text="primary"
          type="outline"
        />
      </>,
    )

    const buttonPrimary = getByLabelText('button-primary')
    const buttonPrimaryStyle: ViewStyle = buttonPrimary.props.style
    expect(buttonPrimaryStyle.borderWidth).toEqual(StyleSheet.hairlineWidth)

    const buttonPrimary2 = getByLabelText('button-primary2')
    const buttonPrimaryStyle2: ViewStyle = buttonPrimary2.props.style
    expect(buttonPrimaryStyle2.borderWidth).toEqual(1)
    expect(getAllByRole('button').length).toEqual(2)
  })

  it('onPress', () => {
    const buttonOnPress = jest.fn()
    const { getByLabelText } = customRender(
      <Button
        accessibilityLabel="button-primary"
        text="primary"
        onPress={buttonOnPress}
      />,
    )

    const buttonPrimary = getByLabelText('button-primary')
    fireEvent(buttonPrimary.children[0] as ReactTestInstance, 'press')
    expect(buttonOnPress).toHaveBeenCalled()
  })

  it('onPress:disabled', () => {
    const buttonOnPress = jest.fn()
    const { getByLabelText } = customRender(
      <Button
        accessibilityLabel="button-primary"
        text="primary"
        onPress={buttonOnPress}
        disabled
      />,
    )

    const buttonPrimary = getByLabelText('button-primary')
    fireEvent(buttonPrimary.children[0] as ReactTestInstance, 'press')
    expect(buttonOnPress).not.toHaveBeenCalled()
  })

  it('onPress:loading', () => {
    const buttonOnPress = jest.fn()
    const { getByLabelText, getByText } = customRender(
      <>
        <Button
          accessibilityLabel="button-primary"
          text="primary"
          onPress={buttonOnPress}
          loading
        />
        <Button
          accessibilityLabel="button-primary2"
          text="primary"
          loading
          loadingText="哎哟哟"
        />
      </>,
    )

    const buttonPrimary = getByLabelText('button-primary')
    fireEvent(buttonPrimary.children[0] as ReactTestInstance, 'press')
    expect(buttonOnPress).not.toHaveBeenCalled()

    expect(getByText('哎哟哟')).not.toBeNull()
  })

  it('square&round', () => {
    const { getByLabelText } = customRender(
      <>
        <Button accessibilityLabel="button-primary" text="primary" square />
        <Button accessibilityLabel="button-primary2" text="primary" round />
      </>,
    )

    const buttonPrimary = getByLabelText('button-primary')
    const buttonPrimaryStyle: ViewStyle = buttonPrimary.props.style
    expect(buttonPrimaryStyle.borderRadius).toEqual(0)

    const buttonPrimary2 = getByLabelText('button-primary2')
    const buttonPrimaryStyle2: ViewStyle = buttonPrimary2.props.style
    expect(buttonPrimaryStyle2.borderRadius).toEqual(9999)
  })

  it('color', () => {
    const { getByLabelText } = customRender(
      <>
        <Button accessibilityLabel="button-primary" text="primary" />
        <Button accessibilityLabel="button-#000" text="primary" color="#000" />
        <Button accessibilityLabel="button-ghost" type="ghost" text="primary" />
        <Button type="link">primary</Button>
        <Button type="hazy" text="primary" />
        <Button type="outline" text="primary" />
      </>,
    )

    const findFirstChild = (b: ReactTestInstance, i = 0) =>
      b.children[i] as ReactTestInstance
    const findText = (b: ReactTestInstance) =>
      findFirstChild(findFirstChild(findFirstChild(findFirstChild(b))))

    const buttonPrimary = getByLabelText('button-primary')
    const buttonPrimaryStyle: ViewStyle = buttonPrimary.props.style
    const buttonPrimaryText = findText(buttonPrimary)

    expect(buttonPrimaryStyle.backgroundColor).toEqual(TOKENS.brand_6)
    expect(buttonPrimaryStyle.borderColor).toEqual(TOKENS.brand_6)
    expect(buttonPrimaryText.props.style.color).toEqual(TOKENS.white)

    const button000 = getByLabelText('button-#000')
    const button000Style: ViewStyle = button000.props.style
    const button000Text = findText(button000)

    expect(button000Style.backgroundColor).toEqual('#000')
    expect(button000Style.borderColor).toEqual('#000')
    expect(button000Text.props.style.color).toEqual(TOKENS.white)

    const buttonGhost = getByLabelText('button-ghost')
    const buttonGhostStyle: ViewStyle = buttonGhost.props.style
    const buttonGhostText = findText(buttonGhost)

    expect(buttonGhostStyle.backgroundColor).toEqual('transparent')
    expect(buttonGhostStyle.borderColor).toEqual(TOKENS.brand_6)
    expect(buttonGhostText.props.style.color).toEqual(TOKENS.brand_6)
  })
})
