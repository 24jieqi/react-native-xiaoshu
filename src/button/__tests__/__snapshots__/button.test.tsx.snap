// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button render snapshot 1`] = `
<View
  accessibilityLabel="button"
  accessibilityRole="button"
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": false,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={false}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "alignItems": "center",
      "backgroundColor": "#0065FE",
      "borderColor": "#0065FE",
      "borderRadius": 4,
      "borderStyle": "solid",
      "borderWidth": 0,
      "flexDirection": "column",
      "height": 44,
      "justifyContent": "center",
      "opacity": 1,
      "paddingHorizontal": 8,
    }
  }
>
  <View
    style={
      [
        {
          "alignItems": "center",
          "flexDirection": "row",
          "flexWrap": "nowrap",
          "justifyContent": "center",
        },
        undefined,
      ]
    }
  >
    <Text
      numberOfLines={1}
      style={
        {
          "color": "#FFFFFF",
          "fontSize": 16,
          "marginLeft": 0,
          "textAlign": "center",
        }
      }
    >
      primary
    </Text>
  </View>
</View>
`;
